### ISSUE TITLE
The ENCORE_RUNTIME_LIB environment variable is not set

### ISSUE THREAD

**Issue by rtritto** [encoredev/encore#1611]
### Reproduction
> repository=[repro-encore-win](https://github.com/rtritto/repro-encore-win) ~ branch=[missing-env](https://github.com/rtritto/repro-encore-win/tree/missing-env)
- `git clone https://github.com/rtritto/repro-encore-win.git`
- `cd repro-encore-win`
- `git checkout missing-env`
- `yarn`
- `yarn start` OR `node ./start.mjs`
  ```console
  C:\Users\<USER>\AppData\Local\Yarn\Berry\cache\encore.dev-npm-1.44.6-3d86b75c4d-10c0.zip\node_modules\encore.dev\dist\internal\runtime\napi\napi.cjs:7
    throw new Error(
          ^
  
  Error: The ENCORE_RUNTIME_LIB environment variable is not set. It must be set to the path of the Encore runtime library ('encore-runtime.node').
      at Object.<anonymous> (C:\Users\<USER>\AppData\Local\Yarn\Berry\cache\encore.dev-npm-1.44.6-3d86b75c4d-10c0.zip\node_modules\encore.dev\dist\internal\runtime\napi\napi.cjs:7:9)  
      at Module._compile (node:internal/modules/cjs/loader:1546:14)
      at Module._extensions..js (node:internal/modules/cjs/loader:1691:10)
      at require$$0.Module._extensions..js (C:\repro-encore-win\.pnp.cjs:9800:33)
      at Module.load (node:internal/modules/cjs/loader:1317:32)
      at Module._load (node:internal/modules/cjs/loader:1127:12)
      at require$$0.Module._load (C:\repro-encore-win\.pnp.cjs:9648:31)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:217:24)
      at cjsLoader (node:internal/modules/esm/translators:329:5
  ```

### Info
OS: Windows 11
Node: v22
encore.dev: v1.44.6 (same behavior with v1.44.5)
encode installed v1.44.5
yarn berry with PnP (Plug'n'Play)

**Comment by rtritto** [comment]
I fixed with env var: `ENCORE_RUNTIME_LIB='C:\\Users\\<USER>\\.encore\\runtimes\\js\\encore-runtime.node'`

---

The load of `encore-runtime.node` binary can be improved:
- remove references of env var `ENCORE_RUNTIME_LIB`
- the install of Encore ([install-the-encore-cli](https://encore.dev/docs/ts/install#install-the-encore-cli)) should create the env var `ENCORE_PATH='C:\Users\<USER>\.encore'`
- the JS `encore.dev` should get the correct path of `encore-runtime.node`: `ENCORE_PATH` + `\runtimes\js\encore-runtime.nod`

**Comment by rtritto** [comment]
Hi @fredr, this issue is related to JS environment (the start of the server is a separated environment)

**Comment by fredr** [comment]
@rtritto could you explain a bit what you are trying to solve? In what scenario do you want to run it this way?

**Comment by eandre** [comment]
I still don't understand what you're trying to do; under what circumstance is the `ENCORE_RUNTIME_LIB` variable not set today? The encore daemon automatically sets it whenever you're running `encore run`, and it's being embedded into any Docker images built via `encore build docker`. So what are you trying to do, exactly?

**Comment by eandre** [comment]
It is not designed to be used like that, no. Due to how Encore's request decoding/response encoding works, it requires the pre-computed API schemas to be provided ("the metadata"). This must be provided via `ENCORE_APP_META` or via `ENCORE_APP_META_PATH`.

What's more, all of Encore's infrastructure resources additionally require runtime configuration to be provided, so that Encore's runtime knows where the infrastructure is located, what implementation to use, how to authenticate, etc.

In short, setting `ENCORE_RUNTIME_LIB` is far from sufficient. That's why Encore provides `encore run`, which takes care of setting up the infrastructure the application requires, computes the metadata and propagates it to the application, etc.

**Comment by rtritto** [comment]
Ok, thanks for the infos.

So if `ENCORE_RUNTIME_LIB` was chosen to be set with `encore run`, imo, this issue can be closed.

In the JS [encore.dev](https://www.npmjs.com/package/encore.dev) dependency, maybe it's better to add an error message. Eg: `Missing "ENCORE_RUNTIME_LIB" env var. Start "encore run" first`.

**Comment by ImBIOS** [comment]
this also happen when I'm trying to use encore with drizzle and try to seed the database in encore-ts

**Comment by mmtftr** [comment]
I'd like a shell environment with encore environment variables set. I occasionally need to run some scripts that import code that uses `encore.dev/log`, and I get this error. It's quite a common need to share code between different projects. I don't want to abstract my logging logic away to this extent.

**Comment by etxdev** [comment]
@ImBIOS Have you found a better solution yet?

**Comment by blackirishmec** [comment]
#### I am running into this issue when using Encore.ts + Prisma w/ DB Seeding. 

### Versioning:
```json
"dependencies": {
    "@prisma/client": "^5.22.0",
    "encore.dev": "^1.46.10"
},
"devDependencies": {
    "prisma": "^5.22.0",
    "tsx": "^4.19.3",
    "typescript": "^5.2.2",
}
```
IDE: VS Code

### The Issue:
If I define my database in database.ts similarly to [the encore.ts prisma example](https://github.com/encoredev/examples/tree/main/ts/prisma):
```typescript
import { SQLDatabase } from 'encore.dev/storage/sqldb';

import { PrismaClient } from '@prisma/client';

const DB = new SQLDatabase('my-db', {
    migrations: {
        path: './prisma/migrations',
        source: 'prisma',
    },
});

const prisma = new PrismaClient({
    datasources: {
        db: {
            url: DB.connectionString,
        },
    },
});

export { prisma };
```
... and then setup seeding via the [prisma seed docs](https://www.prisma.io/docs/orm/prisma-migrate/workflows/seeding) in a `primsa/seed.ts` file, I run into problems:
1. If I follow the docs exactly and setup my `prisma/seed.ts` file using:
    ```typescript
    import { PrismaClient } from '@prisma/client'
    const prisma = new PrismaClient()
    ```
    ... I successfully seed the db with `npx prisma db seed` ([using tsx instead of ts-node - prisma#20369](https://github.com/prisma/prisma/discussions/20369#discussion-5435998)), **however** this seeded data does not make it into my Encore db (tested via `encore db shell...`). I can *confirm* that the seeded data make it into **a** 
prisma db via `npx prisma studio`, however it's a seperate db instance than the one setup for the Encore project.
2. If instead I  setup my `prisma/seed.ts` file using the `prisma` PrismaClient instance exported from `database.ts` (which seems logical in order to link to the PrismaClient which tied to the Encore db):
    ```typescript
    import { prisma } from './database';
    ``` 
    ... I get this error when I run `npx prisma db seed` ([using tsx instead of ts-node - prisma#20369](https://github.com/prisma/prisma/discussions/20369#discussion-5435998)):
    ```zsh
    Running seed command `tsx prisma/seed.ts` ...
    {PROJECT_ROOT}/node_modules/encore.dev/dist/internal/runtime/napi/napi.cjs:7
      throw new Error(
            ^
    
    Error: The ENCORE_RUNTIME_LIB environment variable is not set. It must be set to the path of the Encore runtime library ('encore-runtime.node').
        at Object.<anonymous> ({PROJECT_ROOT}/node_modules/encore.dev/dist/internal/runtime/napi/napi.cjs:7:9)
        at Module._compile (node:internal/modules/cjs/loader:1358:14)
        at Object.transformer ({PROJECT_ROOT}/node_modules/tsx/dist/register-DCnOAxY2.cjs:2:1186)
        at Module.load (node:internal/modules/cjs/loader:1208:32)
        at Module._load (node:internal/modules/cjs/loader:1024:12)
        at cjsLoader (node:internal/modules/esm/translators:348:17)
        at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:297:7)
        at ModuleJob.run (node:internal/modules/esm/module_job:222:25)
        at async ModuleLoader.import (node:internal/modules/esm/loader:316:24)
        at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:123:5)

    Node.js v20.16.0

    An error occurred while running the seed command:
    Error: Command failed with exit code 1: tsx prisma/seed.ts
    ```

### Related Comments in Issue:
- I believe I'm in the same universe as [ImBios's Comment](https://github.com/encoredev/encore/issues/1611#issuecomment-2635804113), however I don't want to have to use the drizzle template to get my own project working, and I don't understand what they mean by:
    > The workaround for the seeding, is I make an internal endpoint for this purpose.

### Workaround:
I've wrastled with this all weekend and the only way to seed my db for now is to include the seeding logic (`main()` definition and `main()` call **inside the `database.ts` file** using the instance of prisma created there (aligning with option *2.* above). This removes my ability to seed via the CLI, and required me to pull the:
```json
"prisma": {
    "seed": "tsx prisma/seed.ts"
}
```
... block out of my `package.json`. 

If anyone has any advice on a fix for this please let me know!

**Comment by fredr** [comment]
We've added a encore exec command that will setup the encore infrastructure and environment variables for your app, and then execute a command.

E.g like this

```
encore exec -- npx prisma db seed
```

Its available in encore version 1.46.16

Let me know if this works for your usecases

**Comment by blackirishmec** [comment]
> We've added a encore exec command that will setup the encore infrastructure and environment variables for your app, and then execute a command.
> 
> E.g like this
> 
> ```
> encore exec -- npx prisma db seed
> ```
> 
> Its available in encore version 1.46.16
> 
> Let me know if this works for your usecases

Fantastic! Thank you *so* much!

**Comment by blackirishmec** [comment]
For existing projects is there a fix? I just ran into another trigger for this error during local testing:
```bash
npm run test

> encore-prisma-example@0.0.1 test
> vitest


 DEV  v3.1.1 /Users/{project root directory}

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  aqi/aqi.controller.test.ts [ aqi/aqi.controller.test.ts ]
Error: The ENCORE_RUNTIME_LIB environment variable is not set. It must be set to the path of the Encore runtime library ('encore-runtime.node').
 ❯ Object.<anonymous> node_modules/encore.dev/dist/internal/runtime/napi/napi.cjs:7:9

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: The ENCORE_RUNTIME_LIB environment variable is not set. It must be set to the path of the Encore runtime library ('encore-runtime.node').
 ❯ Object.<anonymous> node_modules/encore.dev/dist/internal/runtime/napi/napi.cjs:7:9
 ❯ Module._compile node:internal/modules/cjs/loader:1358:14
 ❯ Object.Module._extensions..js node:internal/modules/cjs/loader:1416:10
 ❯ Module.load node:internal/modules/cjs/loader:1208:32
 ❯ Function.Module._load node:internal/modules/cjs/loader:1024:12
 ❯ cjsLoader node:internal/modules/esm/translators:348:17
 ❯ ModuleWrap.<anonymous> node:internal/modules/esm/translators:297:7
 ❯ ModuleJob.run node:internal/modules/esm/module_job:222:25
 ❯ ModuleLoader.import node:internal/modules/esm/loader:316:24
 ❯ VitestExecutor.interopedImport node_modules/vite-node/dist/client.mjs:365:26

This error originated in "aqi/aqi.controller.test.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯


 Test Files  1 failed (1)
      Tests  no tests
     Errors  1 error
   Start at  11:41:05
   Duration  192ms (transform 26ms, setup 0ms, collect 0ms, tests 0ms, environment 0ms, prepare 40ms)

 FAIL  Tests failed. Watching for file changes...
       press h to show help, press q to quit
```

**Comment by fredr** [comment]
run tests via `encore test` instead of `npm run test`. That will also setup your environment correctly

**Comment by fredr** [comment]
I'll close this issue now that the exec command exist. If there is any issues still, write to us on discord or open new issues here.
